### ISSUE TITLE
SpotBugs misunderstands resource management when using mock objects

### ISSUE THREAD

**Issue by ChristopherSchultz** [spotbugs/spotbugs#1489]
Best description is with an example, using EasyMock. Consider the following code:

```
IMocksControl control = EasyMock.createControl();
Connection conn = control.createMock(Connection.class);
PreparedStatement ps = control.createMock(PreparedStatement.class);

EasyMock.expect(conn.prepareStatement(eq("SELECT 1 FROM DUAL"))).andReturn(ps);
```

SpotBugs considers the call to `conn.prepareStatement` to be something which directly produces a `PreparedStatement` object which needs to be cleaned-up. The bug type is:

`OBL_UNSATISFIED_OBLIGATION_EXCEPTION_EDGE: Method may fail to clean up stream or resource on checked exception`

It's probably impossible for SpotBugs to "know" that this is not the case, but in a well-tested code-base, disabling each of these checks will be quite onerous.

Perhaps there is a way to white-list all uses of EasyMock (and probably other common testing frameworks) such that they don't trigger this particular bug detector?

(I'm giving this feedback because the bug description specifically requests it. Keep up the great work!)

**Comment by ThrawnCA** [comment]
@ChristopherSchultz A few of our projects have exclusion rules for test code similar to:

    <Match>
        <Class name="~au\.gov\.qld\.ssq\..*(Test|IT)([$][0-9]+)?"/>
        <Bug category="PERFORMANCE" pattern="UWF_FIELD_NOT_INITIALIZED_IN_CONSTRUCTOR,SEC_SIDE_EFFECT_CONSTRUCTOR"/>
    </Match>

You could probably add OBL_UNSATISFIED_OBLIGATION_EXCEPTION_EDGE as a pattern or OBL as a category. (And update the package specification to match your own source tree, of course.)

**Comment by gtoison** [comment]
I'll close this since there was no further question
